src/lightning/pytorch/cli.py

```diff
@@ -755,3 +777,33 @@ def _get_short_description(component: object) -> Optional[str]:
         return docstring.short_description
     except (ValueError, docstring_parser.ParseError) as ex:
         rank_zero_warn(f"Failed parsing docstring for {component}: {ex}")
+
+
+def _get_module_type(value: Union[Callable, type]) -> type:
+    if callable(value) and not isinstance(value, type):
+        return inspect.signature(value).return_annotation
+    return value
+
+
+class _InstantiatorFn:
+    def __init__(self, cli: LightningCLI, key: str) -> None:
+        self.cli = cli
+        self.key = key
+
+    def __call__(self, class_type: Type[ModuleType], *args: Any, **kwargs: Any) -> ModuleType:
+        with _given_hyperparameters_context(
+            hparams=self.cli.config_dump.get(self.key, {}),
+            instantiator="lightning.pytorch.cli.instantiate_module",
```

Comment:
@carmocca @awaelchli @Borda I changed to this to avoid providing an instantiator during the call to `load_from_checkpoint`. The tests run fine, but now I am wondering what happens when the `pytorch-lightning` package is installed instead of `lightning`. This line could be changed so that the import path is different for each package. But then, how would this be tested? Also, if someone saves a checkpoint using `pytorch-lightning`, should the load work when using `lightning`?

Reply:
We have a CI step where `lightning.pytorch` gets rewritten to `pytorch_lightning` for the CI jobs that build `pytorch_lightning`. If tests pass (they did) it should be a reliable signal that this worked as expected.

> Also, if someone saves a checkpoint using pytorch-lightning, should the load work when using lightning?

However, this most likely doesn't work, but mixing the two packages is strongly discouraged. I guess it could be a problem for different users who share checkpoints.

Reply:
If it becomes an issue, we could write a checkpoint migration that will rewrite the instantiator import path on the fly:
https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/utilities/migration/migration.py

Checkpoint migrations are applied automatically when lightning loads a checkpoint.

Reply:
Migrations only go one way `pytorch-lightning` -> `lightning`, or is it bidirectional?

Reply:
Migrations are just one way for us to define transformations to "upgrade" checkpoints from one version of lightning to another, for example when breaking changes were made so people can still load their old checkpoints. 

It could also be used to define a transformation that replaces the instantiator string if the user is running from within the other package.

Reply:
Is it known how common is it for people to share checkpoints and use them indistinctly in `pytorch-lightning` or `lightning`? If it is not common or it is not known how common, then we could leave it like this. And do an improvement when someone raises the issue. Or should this be considered now before the feature included in a release?

Reply:
I'm ok leaving it like this for now.